platform-intel.c

```diff
@@ -669,14 +724,251 @@ static int read_efi_variable(void *buffer, ssize_t buf_size,
+		       file_name, strerror(errno));
+		return MDADM_STATUS_ERROR;
+	}
+
+	return MDADM_STATUS_SUCCESS;
+}
+
+/**
+ * verify_uefi_acpi_table_length() - verify if ACPI UEFI table have correct length with focus at
+ * OROM.
+ * @table: struct with UEFI table.
+ *
+ * Verify if ACPI UEFI table have correct length with focus at OROM. Make sure that the file is
+ * correct and contains the appropriate length data based on the length of the OROM.
+ *
+ * Return: %true if length is correct, %false otherwise.
+ */
+bool verify_uefi_acpi_table_length(struct uefi_acpi_table table)
+{
+	if (table.length < ACPI_UEFI_DATA_OFFSET)
+		return false;
+
+	if (table.length <= table.dataOffset)
+		return false;
+
+	if (table.length - table.dataOffset != sizeof(struct imsm_orom))
+		return false;
+	return true;
+}
+
+/**
+ * find_orom_in_acpi_uefi_tables() - find OROM in UEFI ACPI tables based on requested controllers.
+ * @uefi_controllers: array of requested UEFI controllers.
+ * @uefi_controllers_number: number of UEFI controllers.
+ * @orom: OROM struct buffer to fill out.
+ *
+ * Find OROM in UEFI ACPI table based on requested controllers.
+ * If found, the buffer with the OROM structure will be filled.
+ *
+ * Return: %MDADM_STATUS_SUCCESS on success, %MDADM_STATUS_ERROR otherwise.
+ */
+mdadm_status_t
+find_orom_in_acpi_uefi_tables(uefi_controller_t *uefi_controllers, int uefi_controllers_number,
+			      struct imsm_orom *orom)
+{
+	mdadm_status_t status = MDADM_STATUS_ERROR;
+	uefi_acpi_table_with_orom_t uefi_table;
+	char path[PATH_MAX];
+	struct dirent *ent;
+	int fd = -1;
+	DIR *dir;
+
+	dir = opendir(ACPI_TABLES_PATH);
+	if (!dir)
+		return MDADM_STATUS_ERROR;
+
+	for (ent = readdir(dir); ent; ent = readdir(dir)) {
+		close_fd(&fd);
+
+		/* Check if file is a UEFI table */
+		if (strncmp(ent->d_name, ACPI_UEFI_TABLE_BASE_NAME,
+			    strlen(ACPI_UEFI_TABLE_BASE_NAME)) != 0)
+			continue;
+
+		snprintf(path, PATH_MAX, "%s/%s", ACPI_TABLES_PATH, ent->d_name);
+
+		fd = open(path, O_RDONLY);
+		if (!is_fd_valid(fd)) {
+			pr_err("Fail to open ACPI UEFI table file. File: %s, Error: %s\n",
+			       ent->d_name, strerror(errno));
+			continue;
+		}
+
+		if (read_uefi_acpi_table(fd, ent->d_name, &uefi_table.table))
+			continue;
+
+		if (!acpi_is_requested_uefi_controller(uefi_controllers, uefi_controllers_number,
+						       uefi_table.table))
+			continue;
+
+		if (!verify_uefi_acpi_table_length(uefi_table.table))
+			continue;
+
+		if (read_uefi_acpi_orom_data(fd, ent->d_name, &uefi_table)) {
+			continue;
+		} else {
+			memcpy(orom, &uefi_table.orom, sizeof(uefi_table.orom));
+			status = MDADM_STATUS_SUCCESS;
+			goto close;
+		}
+	}
+
+close:
+	close_fd(&fd);
+	closedir(dir);
+	return status;
+}
+
+/**
+ * find_imsm_efi_orom() - find OROM for requested controller.
+ * @orom: buffer for OROM.
+ * @controller_type: requested controller type.
+ *
+ * Based on controller type, function first search in EFI vars then in ACPI UEFI tables.
```

Comment:
Please highlight that it is looking for one of the variables defined in arrays , i.e. first match wins.